ENH: add BooleanArray extension array

jorisvandenbossche opened the pull request:
Closes https://github.com/pandas-dev/pandas/issues/21778 
xref closed PRs https://github.com/pandas-dev/pandas/pull/22226 and https://github.com/pandas-dev/pandas/pull/25415, and xref the discussions on missing values in https://github.com/pandas-dev/pandas/issues/28095 and https://github.com/pandas-dev/pandas/issues/28778

This PR adds a BooleanArray extension array ("boolean" dtype):

- It uses the same mask-based approach as the IntegerArray: numpy array for the values + numpy array for the mask
- it does not yet implement any new NA behaviour as being discussed #28778, but for now follows the current NaN behaviour
- It is not yet "produced" from other extension arrays (eg a comparison method on IntegerArray, I would leave that for follow-up PRs)
- There is certainly a lot of code to share between this BooleanArray and IntegerArray (eg in a "BaseMaskedArray" base class), but didn't do that yet here (focusing on the BooleanArray for now)

jorisvandenbossche committed:
ENH: add BooleanArray extension array

jorisvandenbossche commented on pandas/core/arrays/boolean.py:
```diff
@@ -0,0 +1,605 @@
+            data[self._mask] = self._na_value
+            return data
+        else:
+            return self._data
```
I adapted this compared to IntegerArray to return bool dtype if there are no NaNs (this helps to enable boolean indexing automatically for this case). But we should probably make a consistent choice.

TomAugspurger replied:
Yeah, I don't think we want to return object-dtype here if we can avoid it...

jorisvandenbossche replied:
On the other hand, a predictable dtype (not depending on the values in the array) is also nice ...

TomAugspurger replied:
Yeah, as I thought more about it the dtype stability was winning me over. Not sure what's best.

jorisvandenbossche replied:
Switched back to always object dtype for now. There are methods (like astype(bool)) that force you to give bool dtype if possible. We should maybe think about adding a `to_numpy` to our arrays as well (the same for IntegerArray).

jorisvandenbossche commented on pandas/core/arrays/boolean.py:
```diff
@@ -0,0 +1,605 @@
+        data = self._data
+        mask = self._mask
+
+        # coerce to a nan-aware float if needed
```
Still need to update this for Boolean and test this

jorisvandenbossche commented on pandas/tests/arrays/test_boolean.py:
```diff
@@ -0,0 +1,277 @@
+# def test_to_boolean_array_error(values):
+#     # error in converting existing arrays to BooleanArray
+#     with pytest.raises(TypeError):
+#         pd.array(values, dtype="boolean")
```
Currently those actually work, as the coercion routine is using numpy coercion: `np.array(['a'], dtype=bool)` actually works and use bool-ness of the values. 
So the question here is if we want to allow this, or rather be more strict (eg only allow actual bools + integer 0/1)? (personally prefer to be strict)

TomAugspurger replied:
Mmm, as long as actual bools and integer 0/1 as accepted then either works for me.

TomAugspurger replied:
Would be nice to have a decision here, and get tests for it. I think we can justify either, since `np.array([np.nan], dtype=bool)` will be `array([True])`. So we're already deviating from NumPy in this limited case.

Do you have a plan for how to do the actual validation? Will it be somewhat expensive?

What are we allowing in `__setitem__`? Do we convert the value to `bool`, or do we require booleans?

jorisvandenbossche replied:
So an option could be to allow: bool + int/float of 0's and 1's (the float is needed in case it are 0/1's with np.nan), and raise for the rest.

> Do you have a plan for how to do the actual validation? Will it be somewhat expensive?

We can check for actual bool/int/float dtypes (and then for int and float check that the original is equal to the result converted back to numeric; to ensure it were only 0's and 1's). For object we can first do an infer_dtype, and then apply the same logic.

> What are we allowing in __setitem__? Do we convert the value to bool, or do we require booleans?

Currently, we use the same `coerce_to_array` function in `__setitem__` to convert the value(s) to be set to boolean values+mask. So if we restrict like the above, then that follows the same pattern for `__setitem__` (but we could also be more strict in `__setitem__` if that is preferred)

jorisvandenbossche commented on pandas/tests/arrays/test_boolean.py:
```diff
@@ -0,0 +1,277 @@
+
+
+# @pytest.mark.parametrize("ufunc", [np.add])
+# def test_ufuncs_binary(ufunc):
```
still need to implement arithmetic ops

TomAugspurger replied:
What are the expected return types? Would you use IntegerArray for things returning an integer dtype?

ndarray[bool] will sometimes return a float dtype...

```python
In [17]: a = np.array([True, False])

In [18]: a / 1
Out[18]: array([1., 0.])
```

would we return an ndarray there, with NaN for the missing values?

TomAugspurger replied:
I guess returning a PandasArray is also an option... If we want to stay in the algebra of `op(BooleanArray, other)` returning an ExtensionArray

jorisvandenbossche replied:
For now I went with tthe dtype that numpy ops result in, and then if result is bool -> BooleanArray, if is int -> IntegerArray, otherwise if result is float dtype (or something else) return the numpy array.

jbrockmendel commented:
should this be marked as a WIP?

jreback reviewed (changes requested):
is there a reason you did not inherit from IntegeryArray (and dtype) and just copied 

surely these share much more code and could easily share a bae class at the very least

jorisvandenbossche commented:
> is there a reason you did not inherit from IntegerArray (and dtype) and just copied
>
> surely these share much more code and could easily share a bae class at the very least

As I said in the top post: yes this needs to share code with IntegerArray through common base class, but initially focussed on a working BooleanArray and deciding on its behaviour. 
I am happy (and planning) to do that, here or in a follow-up.

I don't think it should directly inherit IntegerArray, but personally prefer a common base class. I think that will make it clearer what is actually shared.

jreback commented:
> > is there a reason you did not inherit from IntegerArray (and dtype) and just copied
> > surely these share much more code and could easily share a bae class at the very least
> 
> As I said in the top post: yes this needs to share code with IntegerArray through common base class, but initially focussed on a working BooleanArray and deciding on its behaviour.
> I am happy (and planning) to do that, here or in a follow-up.
> 
> I don't think it should directly inherit IntegerArray, but personally prefer a common base class. I think that will make it clearer what is actually shared.

i agree a common base class is good and as a follow up is fine

jorisvandenbossche commented:
> should this be marked as a WIP?

@jbrockmendel not in the sense of "this is ready to be reviewed", and it also passing all extension array tests (except one, see below). 
There are certainly still todo's (eg the arithmetic ops and ufunc machinery), but eg some commented out tests are mainly API questions (what behaviour do we want? for which review is welcome). There are also parts that I am not planning to address in this PR (eg adding a scalar pd.NA) but rather in separate PRs.

---

The one failure in the base extension array tests, is for concatting mixed dtypes (`result = pd.concat([df1, df2])`):

```
(Pdb) df1
       A
0   True
1  False
2   True
(Pdb) df2
   A
0  1
1  2
2  3
(Pdb) result
   A
0  1
1  0
2  1
0  1
1  2
2  3
(Pdb) expected
       A
0   True
1  False
2   True
0      1
1      2
2      3
```

So it is expecting to return object dtype. 
However, on master for numpy dtypes, we also coerce booleans to integer:

```
In [23]: df1 = pd.DataFrame({'A': [True, False]})  

In [24]: df2 = pd.DataFrame({'A': [1, 2]})  

In [25]: pd.concat([df1, df2])    
Out[25]: 
   A
0  1
1  0
0  1
1  2
```

What to follow here? (we should probably have some discussion about clearer rules (and document this) for which dtypes coerce with which others. Or does this exist somewhere?)

jorisvandenbossche committed:
enable arithmetic ops + ufuncs

jorisvandenbossche committed:
switch back to object dtype for __array__ + astype tests

jorisvandenbossche commented:
Let me be clearer then: this is ready to be reviewed

jorisvandenbossche committed:
temp

jorisvandenbossche commented:
Are people OK with first reviewing / getting merged this version of a BooleanArray without using the new NA, and only in a follow-up adapt it to use NA ? 
(I prefer that to be able to work on those things in parallel, many aspects of BooleanArray do not involve NA vs np.nan, and those an already be reviewed here).

jreback commented:
yes let’s keep BooleanArray independent and merge first

TomAugspurger commented on pandas/core/arrays/boolean.py:
```diff
@@ -0,0 +1,679 @@
+        if copy:
+            mask = mask.copy()
+    else:
+        if mask is None:
```
Style: flatten this a bit with

```
if mask is None and mask_values is None:
    ...
elif mask is None:
    mask = mask_values
else:
    ...
```

TomAugspurger commented on pandas/core/arrays/boolean.py:
```diff
@@ -0,0 +1,679 @@
+    else:
+        if mask is None:
+            if mask_values is None:
+                # TODO optimize this case?
```
Can you clarify the potential optimization in the comment?

jorisvandenbossche replied:
that's a left-over from copying it from integer, can probably be removed
It might be related to the general optimization that we can do for those mask-based arrays to allow a None mask if there are no missing values.

TomAugspurger commented on pandas/core/arrays/boolean.py:
```diff
@@ -0,0 +1,679 @@
+            else:
+                mask = mask_values
+        else:
+            mask = np.asarray(mask, dtype=bool)
```
hmm when is this case reached? I'm a little confused about the `mask | mask_values` below.

jorisvandenbossche replied:
In principle, you can pass a `mask` manually to this function (and in that case it gets combined with the potential mask of the values). However, this functionality is actually not used right now by BooleanArray.

jorisvandenbossche replied:
I added tests for this (as I think we want to expose this publicly at some point)

TomAugspurger commented on pandas/core/arrays/boolean.py:
```diff
@@ -0,0 +1,679 @@
+            if self._mask[item]:
+                return self.dtype.na_value
+            return self._data[item]
+        return type(self)(self._data[item], self._mask[item])
```
Can you check: does this handle indexing a BooleanArray with another BooleanArray well? Does it have to go through object dtype?

TomAugspurger commented on pandas/core/arrays/boolean.py:
```diff
@@ -0,0 +1,679 @@
+
+    @property
+    def _na_value(self):
+        return np.nan
```
Can this be `self.dtype.na_value`?

TomAugspurger commented on pandas/core/arrays/boolean.py:
```diff
@@ -0,0 +1,679 @@
+        ExtensionArray.argsort
+        """
+        data = self._data.copy()
+        data[self._mask] = data.min() - 1
```
Can avoid the `min` and always use -1?

jorisvandenbossche commented:
> Can you check: does this handle indexing a BooleanArray with another BooleanArray well? Does it have to go through object dtype?

@TomAugspurger boolean indexing does not yet work. Well, it did work when BooleanArray coerced to a bool ndarray, and then it worked automatically (for the case that there are no NAs). 
But now I changed back to always convert to object ndarray, we will need to handle this case specifically. Since that will probably involve adding a check and proper conversion to numpy array in several places in the indexing code, I was planning on doing that later / separate PR.

TomAugspurger commented:
> Since that will probably involve adding a check and proper conversion to numpy array in several places in the indexing code, I was planning on doing that later / separate PR.

SGTM. Have we discussed the API on that anywhere? Do missing values propagate, or are they treated as False?

jbrockmendel commented on pandas/core/dtypes/missing.py:
```diff
@@ -445,7 +445,7 @@ def array_equivalent(left, right, strict_nan=False):
             else:
                 try:
-                    if np.any(left_value != right_value):
+                    if np.any(np.asarray(left_value != right_value)):
```
why is this needed?

jorisvandenbossche replied:
Because "reduce" ufuncs are not yet implemented (and `np.any` seems to use `np.logical_or.reduce` ufunc under the hood)

jbrockmendel replied:
can you add a comment to this effect

jbrockmendel commented on pandas/core/arrays/boolean.py:
```diff
@@ -0,0 +1,679 @@
+
+    @property
+    def _ndarray_values(self) -> np.ndarray:
+        """Internal pandas method for lossy conversion to a NumPy ndarray.
```
newline

jorisvandenbossche commented:
> Have we discussed the API on that anywhere? Do missing values propagate, or are they treated as False?

That's where I want feedback, see the discussion in https://github.com/pandas-dev/pandas/issues/28778

jbrockmendel commented on pandas/core/arrays/boolean.py:
```diff
@@ -0,0 +1,679 @@
+        # may need to fill infs
+        # and mask wraparound
+        if is_float_dtype(result):
+            mask |= (result == np.inf) | (result == -np.inf)
```
This looks like it is copied from IntegerArray, and should not be an inplace operation.  It risks alterting self._mask inplace. xref #27829

jorisvandenbossche replied:
> This looks like it is copied from IntegerArray, and should not be an inplace operation. It risks alterting self._mask inplace. xref #27829

OK, I removed this here altogether, as I don't think we should fill any infs in the BooleanArray division (we should separately also look at this for IntegerArray to see what behaviour we want)

This means that for BooleanArray I follow numpy's behaviour in returning float with inf:

```
In [19]: True / np.array([True, False, True])
/home/joris/miniconda3/envs/dev/bin/ipython:1: RuntimeWarning: divide by zero encountered in true_divide
  #!/home/joris/miniconda3/envs/dev/bin/python
Out[19]: array([ 1., inf,  1.])
```

jorisvandenbossche committed:
Merge remote-tracking branch 'upstream/master' into boolean-EA

jorisvandenbossche committed:
updates for feedback + add BooleanArray docstring

jorisvandenbossche committed:
Merge remote-tracking branch 'upstream/master' into boolean-EA

jorisvandenbossche committed:
try fix test for old numpy

jorisvandenbossche committed:
fix in place modification of mask / follow numpy for division

jorisvandenbossche commented:
@jreback @jbrockmendel @TomAugspurger thanks all for the feedback. Updated based on that, and also fixed the failures (old numpy version 1.13 is behaving differently for subtraction of booleans, for now I just skip the tests for this old numpy version)

TomAugspurger commented:
@jorisvandenbossche are you doing docs here? Would need to update at least

* reference/arrays.rst
* basics.dtypes
* missing_data.rst

TomAugspurger commented on pandas/core/arrays/boolean.py:
```diff
@@ -0,0 +1,725 @@
+    @property
+    def name(self) -> str:
+        """
+        The alias for StringDtype is ``'string'``.
```
string->boolean

TomAugspurger commented on pandas/core/arrays/boolean.py:
```diff
@@ -0,0 +1,725 @@
+
+    @classmethod
+    def construct_from_string(cls, string: str) -> ExtensionDtype:
+        if string == "boolean":
```
I suppose "boolean" -> `cls.name`. I can do the same in StringArray.

jorisvandenbossche committed:
string -> boolean copy paste errors

jorisvandenbossche committed:
add basic docs

jorisvandenbossche commented:
> are you doing docs here? Would need to update at least

Good question: I now added some basic docs (add to the api reference, added to the overview tables of dtypes, added short whatsnew). 
I think it will be easier to write more documentation when we actually implement the new NA behaviour (the current PR still uses np.nan semantics, since there is not yet a pd.NA): docs on how NA in boolean operations behave, how it behaves in indexing, differences with numpy's bool dtype, etc

TomAugspurger commented:
> I think it will be easier to write more documentation when we actually implement the new NA behaviour

Agreed.

jbrockmendel commented:
The diff is showing arrays/boolean.py as unchanged, is anyone else seeing this?

TomAugspurger commented:
> The diff is showing arrays/boolean.py as unchanged, is anyone else seeing this?

Yes... That's strange. You can get click to view the whole file and it's there, but not in the `/files` diff page.

jorisvandenbossche committed:
empty test

jorisvandenbossche commented:
Pushed an empty commit, that seems to have fixed it for me

jorisvandenbossche committed:
fix BooleanDtype construction + doc lint

TomAugspurger commented on pandas/core/arrays/boolean.py:
```diff
@@ -0,0 +1,725 @@
+
+    def astype(self, dtype, copy=True):
+        """
+        Cast to a NumPy array or BooleanArray with 'dtype'.
```
I think this should be updated. We can also cast to IntegerArray.

jorisvandenbossche replied:
> I think this should be updated. We can also cast to IntegerArray.

Updated this, and also added several more tests.

jreback commented on pandas/core/arrays/boolean.py:
```diff
@@ -0,0 +1,725 @@
+    """
+    if isinstance(values, BooleanArray):
+        if mask is not None:
+            raise ValueError("cannot pass mask for BooleanArray input")
```
is this have a test?

jreback commented on pandas/core/arrays/boolean.py:
```diff
@@ -0,0 +1,725 @@
+            values = values.copy()
+    else:
+        values_object = np.asarray(values, dtype=object)
+        mask_values = isna(values)
```
this should be on values_object i think

jreback commented on pandas/core/arrays/boolean.py:
```diff
@@ -0,0 +1,725 @@
+            if mask_values is not None:
+                mask = mask | mask_values
+
+    if not values.ndim == 1:
```
shouldn't these checks be near the top?

are these tested? can you handle ndim==0?

jreback commented on pandas/core/arrays/boolean.py:
```diff
@@ -0,0 +1,725 @@
+            raise TypeError(
+                "mask should be boolean numpy array. Use "
+                "the 'array' function instead"
+            )
```
don't you need an isinstance check on values that it must be a BooleanArray? or ndim==1

jorisvandenbossche replied:
> don't you need an isinstance check on values that it must be a BooleanArray? or ndim==1

There is already a isinstance check for the values being a boolean ndarray (so *not* a BooleanArray). But will add a check for ndim==1

jreback replied:
> There is already a isinstance check for the values being a boolean ndarray (so not a BooleanArray). But will add a check for ndim==1

only when you actually coerce, not here though

this is something we should check (here and in IntegerArray) as if you accidently pass a non ndim==1 then it would be an error (can be a followup)

jorisvandenbossche replied:
> only when you actually coerce, not here though

Not fully sure I understand. We don't do any coercing in this `__init__` constructor, and a few lines above there are isinstance checks that the input can only be boolean ndarrays.

> this is something we should check (here and in IntegerArray) as if you accidently pass a non ndim==1 then it would be an error (can be a followup)

As following your comment, I already added a ndim check; I now raise an error if ndim is not 1 (on the lines below)

jreback commented on pandas/core/arrays/boolean.py:
```diff
@@ -0,0 +1,725 @@
+        return type(self)(self._data[item], self._mask[item])
+
+    def _coerce_to_ndarray(self, force_bool=False):
+        """
```
can you add parameters here & type

jreback commented on pandas/core/arrays/boolean.py:
```diff
@@ -0,0 +1,725 @@
+        coerce to an ndarary of object dtype
+        """
+        if force_bool:
+            if not self.isna().any():
```
is this a cached property? (isna)?

jorisvandenbossche replied:
> is this a cached property? (isna)?

Currently we don't do any caching of any properties (also not on integer array). 
I think it can be interesting to look into that, but would leave that for a separate PR (this can be done for IntegerArray as well)

jreback commented on pandas/core/arrays/boolean.py:
```diff
@@ -0,0 +1,725 @@
+        if dtype is not None:
+            if is_bool_dtype(dtype):
+                return self._coerce_to_ndarray(force_bool=True)
+            arr = self._coerce_to_ndarray()
```
this 2nd path seems duplicative of the below

jorisvandenbossche replied:
> this 2nd path seems duplicative of the below

The difference is that this path still does a `arr.astype(dtype, copy=False)` (to take the passed `dtype` into account)

jreback commented on pandas/core/arrays/boolean.py:
```diff
@@ -0,0 +1,725 @@
+            int_result = int(result)
+            if int_result == result:
+                result = int_result
+
```
what about other names? do we raise?

jorisvandenbossche replied:
> what about other names? do we raise?

No, they work, but the result of the numpy operation already has the correct dtype (eg std, var, .. results in a float), so they don't need to be post-processed.

(but just realizing that min/max should return boolean not an int. For sum and prod the int is correct)

jreback reviewed (changes requested):
you are planning a followup to use a mixin with IntegerArray? there is a *lot* of shared code

jorisvandenbossche commented on pandas/core/arrays/boolean.py:
```diff
@@ -0,0 +1,725 @@
+        return astype_nansafe(data, dtype, copy=None)
+
+    @property
+    def _ndarray_values(self) -> np.ndarray:
```
Note to self: remove this `_ndarray_values` for now since we are not using it?

jorisvandenbossche commented on pandas/core/arrays/boolean.py:
```diff
@@ -0,0 +1,725 @@
+            data[mask] = self._na_value
+
+        op = getattr(nanops, "nan" + name)
+        result = op(data, axis=0, skipna=skipna, mask=mask)
```
```suggestion
        result = op(data, axis=0, skipna=skipna, mask=mask, **kwargs)
```

jorisvandenbossche reviewed (commented):
Thanks for the review! I answered to a few, will make code updates for the others tomorrow.

> you are planning a followup to use a mixin with IntegerArray? there is a lot of shared code

Yes, that's the plan. Although I might first want to do a follow-up to start using pd.NA in BooleanArray, and only then a follow-up to share code with IntegerArray (or if pd.NA is not yet ready, can also switch the order. Although this might also depend on whether we already want to start using pd.NA for IntegerArray or not)

jorisvandenbossche committed:
Merge remote-tracking branch 'upstream/master' into boolean-EA

jorisvandenbossche committed:
add extra tests for constructors + check dimensionality

jorisvandenbossche committed:
validate values when converting to boolean array

jorisvandenbossche committed:
various updates

jorisvandenbossche committed:
fix + test return types of reducers

jorisvandenbossche committed:
fix base reduction tests

jorisvandenbossche commented:
@jreback @TomAugspurger OK, made some substantial changes, recommend to look at the diff of the last commits: https://github.com/pandas-dev/pandas/pull/29555/files/a3e1e931392edd2631c3731a081e4eb16128c80a..90558d696c5407523e9fd0e8204c0a2ac8950653

TomAugspurger commented on pandas/core/arrays/boolean.py:
```diff
@@ -131,11 +131,27 @@ def coerce_to_array(values, mask=None, copy=False):
         values = np.zeros(len(values), dtype=bool)
         values[~mask_values] = values_object[~mask_values].astype(bool)
 
+        # if the values were integer-like, validate it were actually 0/1's
```
Does this validate that they're actually 0 or 1? Don't you also need a check that the min / max is in that bound?

TomAugspurger replied:
Ah, sorry you've astyped to bool, so going back to float will make these 0 and 1.

jreback commented:
have u tested this is indexing? i assume that would be a follow up

jorisvandenbossche commented:
> have u tested this is indexing? i assume that would be a follow up

Yep, see https://github.com/pandas-dev/pandas/pull/29555#issuecomment-553772989

jorisvandenbossche commented:
So to gather the follow-ups:

* Use pd.NA instead of np.nan for missing value indicator + update the logical ops to use the pd.NA behaviour
* Reduce duplication with IntegerArray (implement a BaseMaskedArray base class or something like that). This might depend on the decision on whether we already want to use pd.NA in IntegerArray as well or not.
* Update the behaviour of `any`/`all` reductions with `skipna=False` (https://github.com/pandas-dev/pandas/issues/29686)
* Enable boolean indexing with BooleanArray (we can start with handling the case of no NAs; the behaviour for with NAs still need to be decided: https://github.com/pandas-dev/pandas/issues/28778/)

(will open an issue for this)

jorisvandenbossche commented:
Are you all fine with almost merging this? (or maybe a final round of review?) 
Merging it would make it easier to start working on the follow-ups

jreback commented:
let me look one more round

jorisvandenbossche commented:
Other question: what do we want `pd.array([..])` to return by default (when not specifying a dtype)? 

Currently, you get this behavour:
```
In [9]: pd.array([True, False]) 
Out[9]: 
<PandasArray>
[True, False]
Length: 2, dtype: bool

In [10]: pd.array([True, False, np.nan]) 
Out[10]: 
<PandasArray>
[1.0, 0.0, nan]
Length: 3, dtype: float64

In [12]: pd.array([True, False, pd.NA])
Out[12]: 
<PandasArray>
[True, False, NA]
Length: 3, dtype: object
```

Do we want to return BooleanArray also when there are no missing values (which currently returns boolean PandasArray)? 
Eg for integer, we also do not return the IntegerArray (I think the "rule" now is to basically have the same default behaviour as pandas, so only infer the extensions dtypes that are used by default (datetimetz, period, interval, etc). So if we want to be consistent with that, we should not return BooleanArray.

When a `pd.NA` is present, that's of course something new, and we could in principle decide to already infer the new dtype in that case (although then the inference of the dtype depends on the presence of a missing value or not ..) Although I am not sure this is technically easy to do, as the infer_dtypes does not distinguish different types of NA/NaN at the moment (and this out of scope of this PR, could be a future enhancement).

jreback commented:
I would be +1 on using all of the extension arrays (SA, IA, BA) in ``pa.array`` by default. I dont' see any downside here. I would create an issue to update ``infer_dtype`` to do this, shouldn't be very hard as that can now return IA (with an opt in flag).

jorisvandenbossche commented:
I think our reasoning before was to keep this consistent with the main constructors (eg in Series), so that we could in principle move towards Series(..) being Series(array(..)) at some point.

TomAugspurger commented:
> Do we want to return BooleanArray also when there are no missing values (which currently returns boolean PandasArray)?

I don't think that we should have value-dependent behavior here.

On the broader issue, I'd be fine with `pd.array(..., dtype=None)` starting to infer these new EAs. Will make a new issue for that.

jorisvandenbossche commented:
@jreback ping

jreback commented on doc/source/whatsnew/v1.0.0.rst:
```diff
@@ -102,6 +102,30 @@ String accessor methods returning integers will return a value with :class:`Int6
+
+We've added :class:`BooleanDtype` / :class:`~arrays.BooleanArray`, an extension
+type dedicated to boolean data that can hold missing values. With the default
+``'bool`` data type based on a numpy bool array, the column can only hold
```
do we have a dedicated doc section that you can link?

jorisvandenbossche replied:
> do we have a dedicated doc section that you can link?

Going to write one if we add the actual NA behaviour to boolean array (as this is the actual different with bool dtype that needs to be explained)

jreback commented on pandas/core/arrays/boolean.py:
```diff
@@ -0,0 +1,745 @@
+        return True
+
+
+def coerce_to_array(values, mask=None, copy=False):
```
can you type these as much as possible

jorisvandenbossche replied:
I am not really sure how to type those (I don't see anything in `pandas._typing` for "list like" ?)

jreback commented on pandas/core/arrays/boolean.py:
```diff
@@ -0,0 +1,745 @@
+        if copy:
+            values = values.copy()
+    else:
+        # TODO conversion from integer/float ndarray can be done more efficiently
```
are you preparing this for a followup? issue?

jorisvandenbossche replied:
I'll open an issue (the same is true for integer array)

jreback commented on pandas/core/arrays/boolean.py:
```diff
@@ -0,0 +1,745 @@
+        inferred_dtype = lib.infer_dtype(values_object, skipna=True)
+        integer_like = ("floating", "integer", "mixed-integer-float")
+        if inferred_dtype not in ("boolean", "empty") + integer_like:
+            raise TypeError("Need to pass bool-like values")
```
tests hit here?

jorisvandenbossche replied:
> tests hit here?

Yes, there is a test that passes all kinds of non-boolean-like values

In general, I ran locally pytest with coverage, and there is 97% coverage for this file. The main non-covered things are some parts of the ufunc related code, and some length mismatch errors in the ops code.

jreback commented on pandas/core/arrays/boolean.py:
```diff
@@ -0,0 +1,745 @@
+                values[~mask_values].astype(float)
+                == values_object[~mask_values].astype(float)
+            ):
+                raise TypeError("Need to pass bool-like values")
```
this hit in a test?

jorisvandenbossche replied:
Yes

jreback commented on pandas/core/arrays/boolean.py:
```diff
@@ -0,0 +1,745 @@
+    if mask is None and mask_values is None:
+        mask = np.zeros(len(values), dtype=bool)
+    elif mask is None:
+        mask = mask_values
```
do we need to copy mask_values for any inputs?

jorisvandenbossche replied:
`mask_values` is the result of `pd.isna(values_object)`, so should always be "its own" object (it's never a passed mask), so no need to copy it.

jreback commented on pandas/core/arrays/boolean.py:
```diff
@@ -0,0 +1,745 @@
+
+    if not values.ndim == 1:
+        raise ValueError("values must be a 1D list-like")
+    if not mask.ndim == 1:
```
test hit all this?

jorisvandenbossche replied:
> test hit all this?

Yes

jreback commented on pandas/core/arrays/boolean.py:
```diff
@@ -0,0 +1,745 @@
+        return self._dtype
+
+    @classmethod
+    def _from_sequence(cls, scalars, dtype=None, copy=False):
```
type if possible

jreback commented on pandas/core/arrays/boolean.py:
```diff
@@ -0,0 +1,745 @@
+    @classmethod
+    def _from_sequence(cls, scalars, dtype=None, copy=False):
+        if dtype:
+            assert dtype == "boolean"
```
do we have tests for this?

jorisvandenbossche replied:
> do we have tests for this?

Do we need that? This is an internal assertion, that should never be raised to the user but is here to help the developer (I don't think we should add tests for those asserts). So the "test" is that this actually never occurs in the tests.

jorisvandenbossche replied:
BTW, I can also leave it out (the assert). For BooleanArray, the `dtype` is not very useful. I think this parameter is mainly used for cases where multiple dtypes are possible per array (eg int64, int32 etc for IntegerArray)

jreback replied:
right i think it actually should *always* be None here, as this is internally passed. however a user might *try* to pass something that is not None (so this should maybe be a ValueError)

jreback commented on pandas/core/arrays/boolean.py:
```diff
@@ -0,0 +1,745 @@
+        for i in range(len(self)):
+            if self._mask[i]:
+                yield self.dtype.na_value
+            else:
```
a lot of these are highly shared with IntegerArray (for followup)

jreback commented on pandas/core/arrays/boolean.py:
```diff
@@ -0,0 +1,745 @@
+
+        return result
+
+    def _maybe_mask_result(self, result, mask, other, op_name):
```
can be generic int he Int/Bool mixin baseclass (followup)

jreback commented on pandas/tests/extension/test_boolean.py:
```diff
@@ -0,0 +1,332 @@
+                "__rmod__",
+            ):
+                # combine keeps boolean type
+                expected = expected.astype("Int8")
```
is this right?

jorisvandenbossche replied:
It's what numpy does, eg:

```
In [17]: np.array([True, False]) ** 2                                                                                                                                                                              
Out[17]: array([1, 0], dtype=int8)
```

So for those ops, I just followed numpy's behaviour with boolean arrays.

jreback commented on pandas/tests/extension/test_boolean.py:
```diff
@@ -0,0 +1,332 @@
+
+
+class TestComparisonOps(base.BaseComparisonOpsTests):
+    def check_opname(self, s, op_name, other, exc=None):
```
do you need these overrides here? (you are already doing it above)

jorisvandenbossche replied:
Yes, it's to override that there is no exception being raised (will add a comment about that).

(above is for the arithmetic ones)

jreback reviewed (approved):
looks fine. some questions, can be done on followup, mostly asking about test coverage. if you want to do a minor pass ok, or can merge.

jorisvandenbossche committed:
Merge remote-tracking branch 'upstream/master' into boolean-EA

jorisvandenbossche reviewed (commented):
Thanks for the review!

jorisvandenbossche committed:
small edits

jreback commented:
thanks @jorisvandenbossche other things can be done as followups.

jbrockmendel commented:
@jorisvandenbossche for those of us who didn't follow the thread in real-time, can you summarize the design decisions that got made in the end?  and any decisions that remain un-decided?

TomAugspurger commented:
@jbrockmendel started at https://github.com/pandas-dev/pandas/issues/29556#issuecomment-558247082

jorisvandenbossche commented:
@jbrockmendel this PR has no NA-related design decisions (it's just adding a BooleanArray that has not yet any specific NA behaviour). All discussions are still in the respective issues (https://github.com/pandas-dev/pandas/issues/28095, https://github.com/pandas-dev/pandas/issues/28778) and in the open PR to add NA (https://github.com/pandas-dev/pandas/pull/29597), and will come in a future PR to add NA to BooleanArray.

jorisvandenbossche commented:
@jreback opened two follow-up issues related to improving the conversion / astype (which currently goes through object dtype where it is not always needed): https://github.com/pandas-dev/pandas/issues/29838 and https://github.com/pandas-dev/pandas/issues/29839
